shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java

```diff
@@ -175,15 +176,9 @@ ImeSyncDeferringInsetsCallback getImeSyncCallback() {
     return imeSyncCallback;
   }
 
-  @NonNull
-  public AndroidKeyProcessor getKeyEventProcessor() {
-    return keyProcessor;
-  }
-
-  public void setKeyEventProcessor(AndroidKeyProcessor processor) {
-    keyProcessor = processor;
+  public void setKeyboardManager(KeyboardManager processor) {
```

Comment:
This feels so wrong. If I understand coorectly, depending on the Android engine we're using, we either have 
* `KeyboardManager` -> `TextInputPlugin` -> `InputConnectionAdapter`; or
* `TextInputPlugin` -> `InputConnectionAdapter` -> `KeyboardManager`.
This makes the entire logic entangled, and I'll guess is the reason why there is an `instanceof` in `handleKeyEvent`.

Can we somehow resolve this, either by merging some logic or splitting classes? I'm not familiar with the system but I'm glad to discuss if it helps.

Reply:
This happens because `KeyboardManger` does not own `TextInputPlugin`, or vice versa, but they depend on each other. `KeyboardManager` clearly shouldn't own `TextInputPlugin`. `TextInputPlugin` can probably own `KeyboardManager` but I'm not sure we want to do that.

Reply:
The `instanceof` exists because the `InputConnection` protocol by itself can not distinguish a redispatched event from regular key events. An alternative to this is to pass the keyboard manager to `InputConnectionAdater` and override its `sendKeyEvent` and let the keyboard manager shortcircuit redispatched events (this is what the previous implementation does).

Reply:
On other platforms, `KeyboardManger` owns `TextInputPlugin`. Is there anything special for Android?

Reply:
Basically on other platforms, `KeyboardManger` owns 3 things:
```
KeyboardManger -> responders
               -> textInputPlugin
               -> redispatchEvent  // A callback
```
which correspond to the 3 "responders" in the comment [here](https://github.com/flutter/engine/pull/25628#discussion_r620727084).

Reply:
Looked deeper into this, it looks like the text input plugin only uses the keyboard manager to create new input connections which need the keyboard manager to send IME key events to. It makes more sense to change the `createInputConnection` method to take a keyboard manager.